Onboarding: Add WebPreview component for theme previewing

joshuatf opened the pull request:
Fixes #2469

Adds a `WebPreview` component to load content into an iframe.  Used to preview themes in the theme step in onboarding.

### Screenshots
<img width="1355" alt="Screen Shot 2019-07-22 at 4 21 21 PM" src="https://user-images.githubusercontent.com/10561050/61617525-725c3800-ac9d-11e9-93d6-3ac35588fa38.png">

### Detailed test instructions:

1. Visit the theme step in onboarding `wp-admin/admin.php?page=wc-admin&path&step=theme`.
2. Click on "Live Demo" for any of the themes.
3.  Make sure the loader is shown when loading and that the theme preview is loaded into the iframe.
4. Check that content is resized based on the device toggle (desktop/tablet/mobile).
5. Make sure that clicking "Choose" selects the theme and completes the onboarding profiler.

joshuatf commented:
Thanks for taking a look @jameskoster!

>I hope there's a way we can hide the bar at the bottom of the preview?

Yes, but this is being handled on the WCCOM side (both the demo URL and bar itself come from WCCOM).  I know we've had some discussion around it, but I'll open an issue there if one does not already exist.

Let me know if you see other areas you'd like to change around the preview component screen.

timmyc commented on client/dashboard/profile-wizard/steps/theme/preview.js:
```diff
@@ -0,0 +1,82 @@
+					<div className="woocommerce-theme-preview__theme-name">
+						{ interpolateComponents( {
+							mixedString: sprintf(
+								__( '{{strong}}%s{{/strong}} developed by WooCommerce', 'woocommerce-admin' ),
```
Minor but if we ever intend on having any Marketplace developed themes, or others not developed by us, we might want to have this label be a prop from the theme itself.

joshuatf replied:
That's a good idea.  I'll probably leave this as is for now since we would need more information from the WCCOM side.  But if we do implement something like this, maybe a `source` field in the API response would be enough.

timmyc commented on packages/components/CHANGELOG.md:
```diff
@@ -5,6 +5,7 @@
 - Add new component `<List />` for displaying interactive list items.
 - Fix z-index issue in `<Chart />` empty message.
 - Added a new `<SimpleSelectControl />` component.
+- Added a new `<WebPreview />` component.
```
🌟

timmyc commented on packages/components/src/web-preview/index.js:
```diff
@@ -0,0 +1,90 @@
+	/**
+	 * Content shown when iframe is still loading.
+	 */
+	loading: PropTypes.node,
```
This prop name made me do a double take - perhaps it is just too close to `isLoading`. What do you think about `loadingContent` or `loadingComponent`?

joshuatf replied:
Agreed.  I originally named it `loadingContent` but that threw an error (reserved name if I remember correctly).  Will update this before merging.

joshuatf commented:
>The one ux thing that felt mildly off to me was the lack of an ".is-selected" state on the device toggles in the upper right.

Definitely feels like something is needed.  I did not see anything in the designs, but maybe I missed something.  @jameskoster any ideas here?

>( side note, that text looks off in those tabs )

Hmm, the tabs are still showing centered for me @timmyc.  Browser/WP/WC version?

jameskoster commented:
Curiously Material does not require or even suggest a selected state for action items in an App Bar. I'll follow up with the Muriel team for clarification on this. 

For now, we should ensure the hover / focus styles are applied (ideally with the ripple on click). You can find examples (and code I hope) for this [here](https://material-ui.com/components/app-bar).

jameskoster commented:
Spoke with the Muriel team, I think for maximum accessibility (based on Gutenberg patterns) the icon / background should be inversely colored when active. I updated the design in Figma to include an example.

timmyc commented:
> Hmm, the tabs are still showing centered for me @timmyc. Browser/WP/WC version?

I'm running Firefox 68.0.1, WP 5.2.2, WC 3.6.5

timmyc commented:
@joshuatf setting this to in progress pending changes from the latest round of feedback.

joshuatf commented:
>I updated the design in Figma to include an example.

Fantastic- thanks @jameskoster!  Added the focus/selected colors.

I held off on adding the ripple effect as I think that's a substantial PR on its own and might be needs to be added directly to newspack-components to avoid a decent amount of code duplication.  I'll open some issues around this.

>I'm running Firefox 68.0.1, WP 5.2.2, WC 3.6.5

I see the same in Firefox.  Thanks for noting that @timmyc - will get an issue opened!

Updated the `loadingContent` prop (seems to work as long as we don't pass to the iframe) and rebased this branch.  This one is ready for another review.

joshuatf commented on packages/components/src/web-preview/index.js:
```diff
@@ -0,0 +1,90 @@
+	/**
+	 * Content shown when iframe is still loading.
+	 */
+	loadingContent: PropTypes.node,
```
Thanks 🤦‍♂

joshuatf commented on client/dashboard/profile-wizard/steps/theme/preview.js:
```diff
@@ -0,0 +1,85 @@
+					<div className="woocommerce-theme-preview__theme-name">
+						{ interpolateComponents( {
+							mixedString: sprintf(
+								__( '{{strong}}%s{{/strong}} developed by WooCommerce', 'woocommerce-admin' ),
```
Good idea- added.

joshuatf commented:
Thanks for the review @justinshreve- updated per your feedback.

>It also looks like the Deli and Bistro themes do not have a demo button -- but they do have a demo_url in the response.

I see them with demo buttons.  Do you have these themes installed?  If so, previewing for installed themes is not available for the time being.

justinshreve commented:
> I see them with demo buttons. Do you have these themes installed? If so, previewing for installed themes is not available for the time being.

I do not have either of these themes installed.

I looked at the CSS and it looks like they are actually getting pushed down below another theme. I removed this `height: 100%` and things worked correctly.

Before:

<img width="1501" alt="Screen Shot 2019-07-26 at 1 45 49 PM" src="https://user-images.githubusercontent.com/689165/61970883-b5731f80-afab-11e9-99ac-abdc984409c3.png">


After:

<img width="1358" alt="Screen Shot 2019-07-26 at 1 44 38 PM" src="https://user-images.githubusercontent.com/689165/61970856-a55b4000-afab-11e9-96f8-6dffeb6650b2.png">

joshuatf commented:
@justinshreve Not quite able to replicate, but can remove that line and replace with `margin-top: auto;` to keep things working.

Updated in https://github.com/woocommerce/woocommerce-admin/pull/2681/commits/52d83e6fe8ed57dcebb428eeb2d74c970ae54993.  Let me know if it looks okay in your browser now.

jameskoster commented:
There's a couple of (hopefully small) things I think we should tweak:

• When the preview is open, instead of a back arrow in the App Bar let’s use an `x` to close out the preview
• Can we increase the space between the “Choose” button and the device switcher to 16px?
• When previewing “desktop” mode, can the preview occupy the full screen real estate beneath the App Bar?

joshuatf commented:
Thanks for the feedback @jameskoster.  Updated those 3 items, can you take another look?

<img width="1162" alt="Screen Shot 2019-07-31 at 4 55 49 PM" src="https://user-images.githubusercontent.com/10561050/62198784-dfb24c00-b3b4-11e9-80d3-3acc882b50ba.png">

justinshreve commented:
Great work Josh! 🚢 it.
